### ISSUE TITLE
StoreEnhancer has incorrect parameter type

### ISSUE THREAD

**Issue by Methuselah96** [reduxjs/redux#3768]
### Prior Issues

Introduced in https://github.com/reduxjs/redux/pull/3524.

### What is the current behavior?

**Code**
```ts
type MyEnhancerType = { enhancedStuff: number };

function createEnhancer(): StoreEnhancer<MyEnhancerType> {
  return (createStore) => <S, A extends Action>(
    reducer: Reducer<S, A>,
    initialState?: PreloadedState<S>
  ): Store<S, A, never, MyEnhancerType> & MyEnhancerType => {
    // This store that I just created does not have the enhancer applied to it yet
    // It also doesn't know what enhancers have been applied to it
    const nextStore = createStore(reducer, initialState);

    // And yet we can access a property that we have yet to add.
    // This should produce a type error, but it doesn't.
    const propertyThatsNotOnStoreYet = nextStore.enhancedStuff;

    // It's not until here that we have an enhanced store
    // And should be able to access that property
    const enhancedStore = enhanceStore(nextStore);
    const correctPropertyAccess = enhancedStore.enhancedStuff;

    return enhancedStore;
  };
}

function enhanceStore<S, A extends Action>(store: Store<S, A, unknown, unknown>): Store<S, A, never, MyEnhancerType> & MyEnhancerType {
  return {
    ...store,
    enhancedStuff: 5,
  }
}

```

<details><summary><b>Output</b></summary>

```ts
function createEnhancer() {
    return (createStore) => (reducer, initialState) => {
        // This store that I just created does not have the enhancor applied to it yet
        // It also doesn't know what enhancors have been applied to it
        const nextStore = createStore(reducer, initialState);
        // And yet we can access a property that we have yet to add.
        const propertyThatsNotOnStoreYet = nextStore.enhancedStuff;
        // It's not until here that we have an enhanced store
        // And should be able to access that property
        const enhancedStore = enhanceStore(nextStore);
        const correctPropertyAccess = enhancedStore.enhancedStuff;
        return enhancedStore;
    };
}
function enhanceStore(store) {
    return Object.assign(Object.assign({}, store), { enhancedStuff: 5, 
        // Ignore this for now. See https://github.com/reduxjs/redux/issues/3767 for more discussion
        replaceReducer(nextReducer) {
            return store.replaceReducer(nextReducer);
        } });
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgQQMY2BAdgGhVgTzQ2zwBFgBnMAQxlQAs8AFKAUwBsIaATNngMow6bPACV+AV1RsoeAKpZKkgEaVUUYCrZwAvnABmUCCDgBydj0kAPMwG4AUA4D0zuIvYc2NSvzgwCMDZKBwCguABRaxg2LEFhGIAeIRE8KJilTCwAPjgAXjgAbXTYyiyAXTg2aNieSiKsNgA3WXKHODgAfjgUmPa4AC4ehJ0AMkiazOxHB2AsGKgDGhlh6DZE-oF8uBpCHH7kKsm6lHQs7eRCYiz9jt62dO3Glrl+x4KEXQdchH6eKlo9AYQwo1DojESyGyjg6AHM2DB7gAKACUQwEMLgyjUGi0bCRHCoGVkQ1R+VyTQgwB4aPcSlU6k02kx7DAHGWbAkVhkUESADk2AB3e5HDInAR4AWC67YerVMX1KFI-odRrRLnSElwDU8-lC+6SoUypTZfq0oRrRIlOL3PXCkZ4e7pbKG6VnWWOkbpNLRXLjdKY1xwZjsgiw4ySOJwHgQYJYMzwQkAax0AgIKggHAAdBA1LImjRtBxOp1+kHCmmQBns7nfFACyovOVUUMAPJ5+uFrzJaEOL6hQKpmBrCJYBi7XXvRC6T0iKfPWS5ArK1Xy9HD9ij8dYHkW9gAYXYdGgVuis5izocKPJq03Y4nsj3bEP3g3p5g54evsc1UgsH8g63g8947o+G7Pkeb5Tp8n7zs0i7bBsdzbLsBC3Cgoq1Iq7pYBcVw4d8K5wJYmpQEMOqyMkeBQuhYCeNwfDxCInRDKwnAMfwtoCKa155LkT7vrUXGwb61EiR+EwwH6kkzGEOgALIEFuD5QAAKoBHxVCBMjxJIBgGEMWCSFWsh6DMBiRjhcAaK+wHbjyLZAcpoG8opzk8upQQ-P07AwJIUC4UiNkiE+vG5FRGHylhpwkDkREdCRPLkVIuoSigLoqnAczABgNAcPcLHBvRvCcSMPZmuulppcgeALnIcBudpsieWw0mNfZzUad5HQdEGqkMFQWLgf447wAAknAABWkiUPAwUxDw0axvUWAQPA44tCNOixPZ0A7GAbLAH4w5ZfABAIplQZjfAeWUBAS1xgmcBJqtgpwIKo1abtUD1BtOjaLE+2Hcd905ZlqCyvAaqIsNBTzWwT5Iolsh4NluX5SMKIzD1cBBpci3nfAgo6Kguw7KgMiUPUNBwHREBBLABAjXQ706H9cCE-4928DwWbg5DtPGAzAT9XQlB8mtrZYE+ACaCJPPKT5ZjtD66fp2M9VdMBmCta1wJGGAcHADCyDoMCfcTxs0JtZMq6Bi2zWsl1uPjWIMBAkgcIt2g7I2ZvcxTwT1ObLN08LBD80o8B2zpT7bDHCPgUi0OhZiHQQ1H1nQOw6CsPTsgBGglP1AUCfxGsytNWrBgawlCL+bhZdPpiuiOP2Fk7rFX0PgJ1WYXE2GxdkSKO+wlXsBFNX61gL0QIKuDT7P8-ZOa4GT7V8H1e1KktW1SlNWpgG-HXfkBYgmVZpfo+iJlTd+fpQwAKz7M7cBjbCq3sCNg0GHtr1Zj0NgbMYAwDAJQAYrhYQ5QYKoLMEMQDOBItYSalBEFSGsM4KgyhgjOAAMwAHYABs+DDB7RAGsaMVBUAzTKNgTKrJ2QyAoryKUIoooDx6K6Y0cpjiKmHplVc6oUpamYXaA0cApTcNNDjVelpWEOgkUaHClAN4vDwNvFyu84DjA0R5I+AjiL1zPtfLMDCOTMOTvKZh14fBATEQoyRyjVEowavvDqh8vLaNce5TqQQ056H6F8L4QA)

### Steps to Reproduce

See code above.

### What is the expected behavior?

Trying to access enhanced properties right after calling `createStore` should result in a type error.

### Environment Details

N/A

**Comment by Methuselah96** [comment]
Partially addressed by https://github.com/reduxjs/redux/pull/3772, but there's more to be done after that PR gets merged.

**Comment by Methuselah96** [comment]
Nevermind, just made https://github.com/reduxjs/redux/pull/3776 as a separate PR that can get merged whether https://github.com/reduxjs/redux/pull/3772 gets merged or not.

**Comment by alestor123** [comment]
If the code works in JS but fails only because of typing in TypeScript, I'd suggest you to use "any" to ignore it:`
const store = () =>
  createStore(
    reducer,
    initialState as any
  )`

**Comment by Methuselah96** [comment]
@alestor123 I'm aware that I can use `any` to ignore basically any problems with types. My goal in creating this issue is to fix the types not just ignore them (otherwise I might as well be using JS). This problem is more extensive than just the single example provided above. I will provide more examples of why the current types are problematic if that would be helpful (for example getting the correct type after composing enhancers is not possible with the current types).

I have also a PR attached to this issue that fixes the problem. Let me know if you think the proposed solution is incorrect.
